### DISCUSSION TITLE
Can my database schema and views viewed by anyone?

### DISCUSSION THREAD

**Discussion by jonathanrose9** [supabase/supabase#12439] [Questions]
Hello guys,

we are using Supabase in a NextJS application and we are very happy with it. All our tables are stored in the (default) public schema. For our server side code we initialize a client that uses the service_role secret key (server side) and for our client side code we initialize another Supabase client that uses the anon public key (client side). Anyone using our app is able to read our Supabase URL and Anon key because we initialize it client side. I know this is not a problem because we can use RLS to secure our data.

For example here [Supabase Schema](https://supabase-schema.vercel.app). You enter the Supabase URL and the Anon Key here and you get all the information about the schema of our database.  Everyone who uses our app knows the Supabase URL and the Anon Key and just need to put it in the app.

I really like the Supabase schema app and the creators did a great job. I don't want to criticize the app at this point. I would just like to understand why by default the anon key is able to query the entire schema. Any person can read the entire data schema and the relationship of the tables to each other. From this anyone can derive sensitive information that we don't really want to share. In addition, I noticed that using the REST API provided by Supabase, by default all views and their contents can be queried using the Anon Key. Do I lack the necessary PostgreSQL experience and did I do something wrong when creating the view or is this a deliberate behavior of Supabase?

Is there a possibility to lock the anon key or at least take away its rights to protect our schema as well as the data in a view which is in the public schema?

I wonder if I have misunderstood something or am missing something. But for developers who have little experience with PostgreSQL databases these default settings can have serious consequences. Thank you very much for your answers and help.

**Comment by GaryAustin1** [comment]
I saw you asking in another thread and that is the way to shut it off.   But due to this fix https://github.com/supabase/supabase/discussions/9314 it is not working at the moment.  If critical you could contact support until they resolve it.

**Reply by DanMossa** [reply]
What is the thread with the answer?

**Comment by mailbox4ashish** [comment]
Hi Jonathan, I have the same question as you and was wondering if you found the answer to your question.

**Comment by macMikey** [comment]
why not stick your sensitive tables behind a schema that is not **public**? in fact, you could have all your tables in a private schema, then have public views that can only access certain columns of those tables (or you could try using CLS, which is in alpha).
supabase is not the only rest-enabled rdbms that works this way. in many cases, the product authors seem to think of this as a feature, not a bug. another option, is to force many/most/all operations to be conducted via functions, either db functions or edge functions.

**Comment by GaryAustin1** [comment]
You can turn off the REST OPEN API endpoint by returning fake info.  https://postgrest.org/en/v12/references/api/openapi.html#overriding-full-openapi-response  Also make sure you run:
```
alter role authenticator
  set pgrst.db_root_spec = 'public.root';
  ``` 

after you add that function...
